### PR TITLE
Fix mono basis GetEuler bug and marshalling/unmarshalling

### PR DESCRIPTION
**Pull request by cart**
This brings the C# Basis implementation in line with the C++ implementation and reverts the transposition when marshaling/unmarshaling.

This makes the Basis functions behave as expected (they were written under the assumption that the values were transposed internally), but it does mean that the internal values (basis.x, basis.y, basis.z) are transposed.  This is the behavior in C++ but not in GDscript, so it should be changed.

I believe this pr is the best short term fix, but we really need a Basis.cs rewrite that stores untransposed values and performs operations under that assumption.  We could then reapply @NathanWarden's transpose operations (which were reverted here).

While we wait, users that need the real x,y,z axis values should use GetAxis(i), where i = 0 for x, i = 1 for y and i =2 for z.

[More context here](https://github.com/godotengine/godot/issues/16937)

### CODE REVIEW

**Comment by hpvb** [comment]
I'm trying to work out how and what to fix for 3.0.z. Thanks for this.

**Comment by aaronfranke** [comment]
Related: https://github.com/godotengine/godot/pull/17046 and https://github.com/godotengine/godot/issues/16937

Likely has some conflicts with: https://github.com/godotengine/godot/pull/17134

**Review by neikeq — APPROVED** [review]

**Comment by neikeq** [comment]
So what is left in order to fix #14553 and #11901 is to make these use GetAxis right?

**Comment by cart** [comment]
Yup!  I'll do that when I get home today.

**Comment by cart** [comment]
Ok I changed x/y/z to properties that use GetAxis and adjusted all basis methods to use _x/_y/_z (or indexers).  I also added SetAxis so the properties are settable.  A few notes/questions:
1) Can we use PascalCase here (and in the other C# classes?  vector.x, Transform.basis, etc)
2) The indexer operators still use the internal values.  ex: basis[0], basis[0, 1]
3) This pr is a good bandaid fix, but I still think that long term it would be better to rewrite Basis.cs (see my other branch [here](https://github.com/cart/godot/tree/mono-basis-rework) ... it doesn't currently work. see #16937 for context).  It makes the implementation simpler / faster and would fix the indexer operators.  If you want to discuss that more, lets continue the conversation there.

**Comment by aaronfranke** [comment]
Before you rewrite or significantly change Basis, we should merge [my PR](https://github.com/godotengine/godot/pull/17134) which converts all mentions of `float` into `real_t` within Basis and other classes. Otherwise we'd have to spend time resolving conflicts.

File: modules/mono/glue/cs_files/Basis.cs
Comment (neikeq) [inline modules/mono/glue/cs_files/Basis.cs]: I'm wondering, is `[0, ...]` supposed to return `x` or `_x`? I mention it because if that is the case and it's fixed in the future, this code would be broken.

Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: Yeah thats a good question.  I *think* @tagcup would say that m[0] should be rows (_x) and not columns (x) based on our conversation in #16937, but ill let him/her chime in.

As this is the same class, we could also use m._x[1] instead of m[0,1]

Comment (neikeq) [reply modules/mono/glue/cs_files/Basis.cs]: I think we should also replace _x, _y and _z with 9 floats that are named accordingly to the row and column they represent, just like [OpenTK's Matrix3](https://github.com/mono/opentk/blob/master/Source/OpenTK/Math/Matrix3.cs#L38-L63).

Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: Thats definitely less confusing and it has my vote too.  But I also vote that we unbreak master with the current change first.  Its been in a bad state for awhile now and I think fixing it is more important than implementation details.

Comment (neikeq) [reply modules/mono/glue/cs_files/Basis.cs]: I agree. There are two conflicts after the last PR I merged. It can be merged after that.

Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: Cool.  It should be good to go now

**Comment by neikeq** [comment]
Thanks! :)

File: modules/mono/glue/cs_files/Basis.cs
Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: Unless the i in m[i][j] refers to rows, you'll get a bunch of confused users and probably PRs with mistakes in matrix indices.

@reduz was saying OpenGL reverses indices in maths so it won't change I'm law here blah blah but it's totally irrelevant because 1) that's not true today (OpenGL accepts both versions) 2) that'd only be relevant if you're going to write a new renderer under drivers/ where godot's matrix is converted into OpenGL's matrix so has nothing to do with ordinary users.

I agree that it's probably best to use mij with i,j=1,2,3 (or 0,1,2) and/or overload a two-index array operator and exclusively use it in all the code.

Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: i does currently refer to rows (see my other comment)

Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: I have no problem with using mij (ex: m01) or something similar.  But R0C1 is unambiguous so I dont have a problem with that either

Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: Mij is also unambiguous unless you skipped high school

Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: or see the top picture on wikipedia for example https://en.wikipedia.org/wiki/Matrix_(mathematics)

Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: What I'm trying to say is, there is no such thing as column-major in written math. It's just a programming language thing regarding how multidimensional array operators access the underlying data in memory. *Naming* a variable Mij has nothing to do with those operators, it'd just be its mathematical name.

Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: and mathematical notation is always row-major

**Comment by tagcup** [comment]
Actually, doesn't C# have multidimensional arrays as a built-in?
Apologies about the confusion BTW.

File: modules/mono/glue/cs_files/Basis.cs
Comment (cart) [reply modules/mono/glue/cs_files/Basis.cs]: I wasn't trying to make a point about mij, just that R0C1 isnt any worse from an information presentation standpoint.  I'm 100% ok with mij.

**Comment by cart** [comment]
Yeah we could easily use those.  I figured we chose vectors so we could use some of the methods they have, but that doesn't appear to be the case.  Maybe its a stack vs heap thing?  Arrays (multidimensional or not) will be on the heap, but Vector3s are structs so they _could_ be on the stack (but thats a clr implementation detail)

**Comment by tagcup** [comment]
Ah I see, that's too bad. I also looked it up and it appears constant array dimensions can't be a part of data type in C#. That's pretty interesting but C# is more like a toy/scripting language running on top of a VM/JIT'ed so that's not too surprising for me (actually, given the amount of money MS has been throwing at C#, it is still a bit surprising). Well at least it does have stack allocation.

**Comment by tagcup** [comment]
So given that there's also no autovectorization with Mono, something like a bunch of Mij fields looks like the most sensible way going forward

**Comment by cart** [comment]
Agreed!

**Comment by hpvb** [comment]
Cherry picked into 3.0.3

**Pull request merged**
